Support quorum tessera blockchain node

rodion-lim-partior opened the pull request:
This PR is raised to add Quorum and Tessera support for `firefly-cli`. We would like to ease the creation of a development environment for networks using the combination of Blockchain and Private Manager mentioned, and in future, build capabilities in Firefly stack such as extending the blockchain connector to cater for additional functionalities exposed by the combination.

This section describes some of the functionalities implemented from both an external and internal command line perspective:

**External ff command line perspective**
Unsupported combinations / algorithms will be explicitly disabled on the command line flags.

Supported Consensus Mechanism (only when Quorum is selected)
- Clique

Unsupported Consensus Mechanism (only when Quorum is selected)
- RAFT (Intend to support in future, needs command line tweaks, different discovery mechanism)
- IBFT (Intend to support in future, needs Genesis file tweaks)
- QBFT (Intend to support in future, needs Genesis file tweaks)

Supported Client Combinations
- Quorum + EthConnect + Tessera
- Quorum + EvmConnect  
- Quorum + EthConnect

Unsupported Client Combinations
- Quorum + EvmConnect + Tessera (Intend to work on this next)
- Besu + EvmConnect + Tessera
- Besu + EthConnect + Tessera

**Internal ff command line perspective**
1. There are many dependencies to spin up a Quorum/Tessera pair, one example is Quorum having to wait for Tessera to be in a ready state before starting up, we implemented separate docker entrypoint scripts to handle these dependencies for both Quorum and Tessera instead of passing docker args to the binary, so that it is easier to extend more complicated configurations in future. These entry point scripts are copied into the docker volumes during the first time setup
2. To spin up a Tessera instance, we need to generate a private/public key pair. When sending private transactions, Quorum/Third party client needs to use the Tessera public key as an identifier to determine the party to send the unencrypted payload to. The keypairs are pre-generated and then copied into the docker volumes during the first time setup. For this implementation, Tessera configuration takes in a path where the keys are stored, and the same configuration is passed to the binary
3. All Tessera keypairs are dynamically generated, so we need to store this in the stack state / persist them down to disk so that applications built on top have a reference point
4. On top of the Tessera keypairs, private payloads are being stored in the Tessera database which is a configuration property. For the dev implementation, we chose to go with a h2 database that is automatically created within the docker volume if it does not exists. There are other ways of storing this data, such as using a Postgresql database, but for the purpose of a development environment, we feel that h2 is sufficient
5. Tessera instances need to constantly poll each other on the p2p port to determine the status of the members that are connected to each other. The list of members and their endpoints are passed in as a configuration file when we call the Tessera binary. This configuration file is generated in the shell script of the docker entry-point file
6. There are three API servers exposed for Tessera, q2t (9101), p2p (9000) and tp (9080). All three servers are on different ports and need to be exposed internally. Third party port is deliberately exposed on the host, since clients outside of the docker network can choose to use an external signer to hit the `/storeraw` endpoint, or require to perform health checks on the `/upcheck` endpoint
7. Based on the number of members passed in to the CLI, that will determine the number of Quorum and Tessera containers that will be added to the docker compose. Each Quorum has it's own docker volume to store it's signer keys. Instead of having all keys for all members in a single Geth containers, each Quorum only has it's own signer keys. The same is being implemented for Tessera. Similar to the default implementation, we unlock the keys within the Quorum node itself
8. EvmConnect needs to point to the correct Quorum node as per previous bullet point, instead of pointing to a single Geth node in the default implementation
9. Discovery mechanism needs to be taken care of since there are multiple Blockchain nodes, this can be done either via `bootnodes` or a static list of quorum connection identifiers. We have implemented all Quorum nodes to use the first member's node as a bootstrap node to discover each other

Resolves https://github.com/hyperledger/firefly-cli/issues/309

peterbroadhurst commented on internal/blockchain/ethereum/private_transaction_manager.go:
```diff
@@ -0,0 +1,262 @@
+		"jdbc": {
+			"username": "sa",
+			"password": "",
+			"url": "jdbc:h2:./${DDIR}/db;TRACE_LEVEL_SYSTEM_OUT=0",
```
Want to confirm that this storage for the simple dev-grade H2 DB is managed equivalently to the storage for all the other components - like the key management, or if there were an SQLite database (rather than PSQL) being used for other components that need an RDBMS.

This is important as the CLI has lifecycle for stop/start etc. that all depend on this.

rodion-lim-partior replied:
Key management is currently implemented as a persistent file inside the docker volume, generated during stack init, and copied over during first time setup.

This db is required to store the Tessera payloads, it is a configurable value, but for purpose of a development environment, we think the default of h2 is good enough, eventually we might extend support to remote databases such as Postgresql.

peterbroadhurst commented on internal/blockchain/ethereum/private_transaction_manager.go:
```diff
@@ -0,0 +1,262 @@
+				"sslConfig": {
+					"tls": "OFF"
+				},
+				"communicationType" : "REST"
```
If the Quorum to Tessera comms is configured via REST, why do we also have UNIX domain socket configuration above?

rodion-lim-partior replied:
Will remove the above check and removal of the Unix domain socket configuration. We intentionally want to set q2t to REST since we made the choice to not share storage between Quorum and Tessera.

peterbroadhurst commented on internal/blockchain/ethereum/private_transaction_manager.go:
```diff
@@ -0,0 +1,262 @@
+	return nil
+}
+
+func CreateQuorumEntrypoint(ctx context.Context, outputDirectory, volumeName, memberIndex string, chainId int) error {
```
Ok - so this isn't just `Tessera` private transaction manager support, but also Quorum support - that's significant and interesting. Really needs to be covered in the description etc.

rodion-lim-partior replied:
Apologize for the mismatch in the naming of the file, starting Quorum has some coupled dependencies to Tessera, and hence some of the variables are common across the two containers.

I'll refactor to make this clearer.

rodion-lim-partior replied:
Shifted private_transaction_manager.go to a Quorum specific package, and refactor out Quorum docker entry point file creation to its own quorum.go file

peterbroadhurst commented on internal/blockchain/ethereum/private_transaction_manager.go:
```diff
@@ -0,0 +1,262 @@
+	export QUORUM_API="clique"
+fi
+
+TESSERA_URL=http://member%stessera
```
This seems to imply Tessera would always be enabled if I chose Quorum as my ethereum blockchain protocol runtime rather than Besu.

I think that is not an appropriate assertion.

Choosing Besu or Quorum as your EVM is one choice.

Choosing whether you have an off-chain private transaction manager on top feels like a separate choice.
I think it's really important to keep those clearly distinguished.

rodion-lim-partior replied:
Agree that Quorum / Besu can run without a private transaction manager. We struggled slightly on this decision because we would strongly recommend that the better alternative is to run Geth / Besu clients instead of using Quorum if the choice is to run without a private transaction manager.

But agree on your point those choices should not be forced onto the user, will refactor and allow Quorum to start without Tessera.

peterbroadhurst reviewed (changes requested):
Thank you @rodion-lim-partior for the contribution here.

I have spent some time doing the first round of code level understanding of the proposal, and you will see it lead to a significant number of questions that are about the architecture and purpose of this PR.

So I have paused my review for you to take my high level feedback on board.

I would summarize it as:
- There are three things in this PR, and they need to be separately choosable and clear on the `ff` CLI options:
    1. Choosing Quorum (IBFT/QBFT/Clique/RAFT) rather than Besu/Geth as the ethereum blockchain runtime for my dev environment
    2. Choosing to enable the Tessera off-chain private transaction manager on top of my EVM base blockchain
    3. Choosing to pass a set of environment variables to all containers in a stack
- I believe at the end of this PR being ready to merge, from reading the code the following combinations should be possible:
       - Quorum + EVMConnect
       - Quorum + EthConnect
       - Quorum + Tessera + EthConnect
- Unless you do significantly more work, I think the following combinations will fail and need to explicitly fail on the `ff` command line
       - Quorum + Tessera + EVMConnect
       - Besu + Tessera + EthConnect
       - Besu + Tessera + EVMConnect
- The comments on the PR ahead of the next review, I would request come spelling out what you have done both from:
    1. An external `ff` command line perspective
    2. An internal CLI perspective on how what you've done fits with what exists (ports, storage, keys etc.)

With the above addressed, it will be much easier for me and other maintainers to understand what is being proposed in a large PR like this. It's very hard for me to review without this guiding outline to review the code against.

I hope that's helpful, and thanks again for what looks like a significant contribution for established networks that have built upon Quorum and/or Tessera and are looking for a development environment setup.

rodion-lim-partior commented on cmd/init.go:
```diff
@@ -271,5 +272,6 @@ func init() {
 	initCmd.PersistentFlags().StringArrayVar(&initOptions.OrgNames, "org-name", []string{}, "Organization name")
 	initCmd.PersistentFlags().StringArrayVar(&initOptions.NodeNames, "node-name", []string{}, "Node name")
 	initCmd.PersistentFlags().BoolVar(&initOptions.RemoteNodeDeploy, "remote-node-deploy", false, "Enable or disable deployment of FireFly contracts on remote nodes")
+	initCmd.PersistentFlags().StringToStringVar(&initOptions.EnvironmentVars, "environment-vars", map[string]string{}, "Common environment variables to set on all containers in FireFly stack")
```
Hi, this flag is more to address this issue https://github.com/hyperledger/firefly-cli/issues/302, but one of the other consideration was having to hardcode the bootnode URL to the first member of the Quorum network, and we were thinking that eventually, we can improve on the flexibility of the tool if we were able to share variables across the containers in a more generic manner.

rodion-lim-partior commented on cmd/init.go:
```diff
@@ -246,6 +246,7 @@ func randomHexString(length int) (string, error) {
 func init() {
 	initCmd.PersistentFlags().IntVarP(&initOptions.FireFlyBasePort, "firefly-base-port", "p", 5000, "Mapped port base of FireFly core API (1 added for each member)")
 	initCmd.PersistentFlags().IntVarP(&initOptions.ServicesBasePort, "services-base-port", "s", 5100, "Mapped port base of services (100 added for each member)")
+	initCmd.PersistentFlags().IntVar(&initOptions.PtmBasePort, "ptm-base-port", 4100, "Mapped port base of private transaction manager (10 added for each member)")
```
This is to expose the Tessera third party API endpoints to other clients residing outside the docker compose network. One use case would be to send an externally signed transaction from an application on the host. We first have to send the unencrypted payload to the `/storeraw` endpoint in Tessera, before calling the `/sendRawPrivateTransaction` endpoint and passing in the signed transaction populated with the Tessera hash in Quorum.

Also, external monitoring apps outside of the bridge network can hit the tessera endpoint to health check the private transaction manager.

EnriqueL8 replied:
I follow the same thought process as Peter which is why not use the service port range for this? Based on your argument there is a single port needed per Tessera instance to be exposed. What is that is was more difficult to allocate a new port in the services ranged, and make sure other services to do use it than just creating an isolated new port range?

rodion-lim-partior commented on internal/blockchain/ethereum/ethereum.go:
```diff
@@ -32,8 +32,10 @@ import (
-	Address    string `json:"address"`
-	PrivateKey string `json:"privateKey"`
+	Address       string `json:"address"`
+	PrivateKey    string `json:"privateKey"`
```
For our use case, only the public Tessera key is necessary. Applications wanting to share private payloads need to know which Tessera to route to.

We weren't too sure why the Blockchain signer keys are exposed in this data structure, so we went to mirror the same structure for Tessera. I'll remove the Tessera private key from this structure.

There are more complicated setup for Tessera where keys need to be instantiated in Hashicorp Vault. For purpose of the development environment, we want to do away with these complexities as well.

peterbroadhurst replied:
Cool - that's great @rodion-lim-partior 

That structure is used for building the `stack.json` file - when I (or @EnriqueL8) come back round on this one, I think we need to do a bit of digging into the existing use to see why `privateKey` is there at all already. It might just be for reference.

The `CreateAccount` in the signer is the main place this is updated, when a new account is added - but that's after it's copied to the volume.

https://github.com/hyperledger/firefly-cli/blob/1378d838891cd0a43d9a255dac9b707a69f13c59/internal/blockchain/ethereum/ethsigner/ethsigner.go#L207-L210

rodion-lim-partior commented on internal/blockchain/ethereum/private_transaction_manager.go:
```diff
@@ -0,0 +1,262 @@
+	}
+	peerList := strings.TrimSuffix(sb.String(), ",")
+	content := fmt.Sprintf(`export JAVA_OPTS="-Xms128M -Xmx128M"
+DDIR=/data/qdata/dd
```
removing duplicates, we use REST to communicate between Quorum and Tessera, we will do another round of refactoring when we support Besu in future, right now, we explicitly disable Besu with Tessera.

rodion-lim-partior commented:
@peterbroadhurst, I addressed your review comments and refactored the code. Also left a high level overview in the description of this PR, hope this helps to give a clearer picture on the changes that are being proposed.

EnriqueL8 commented on cmd/init.go:
```diff
@@ -200,6 +206,33 @@ func validateBlockchainProvider(providerString, nodeString string) error {
+	}
+
+	if v != types.QuorumConsensusClique {
+		return errors.New("support for raft/ibft/qbft will come in future")
```
```suggestion
		return errors.New("consensus algorithms such as raft/ibft/qbft for quorum not supported")
```

rodion-lim-partior replied:
Raised the following issues:
1. https://github.com/hyperledger/firefly-cli/issues/312
2. https://github.com/hyperledger/firefly-cli/issues/311

EnriqueL8 commented on internal/blockchain/ethereum/ethereum.go:
```diff
@@ -32,8 +32,9 @@ import (
-	PrivateKey string `json:"privateKey"`
+	Address      string `json:"address"`
+	PrivateKey   string `json:"privateKey"`
+	PtmPublicKey string `json:"ptmPublicKey"`
```
```suggestion
	PtmPublicKey string `json:"ptmPublicKey"` // Private Transaction Key used for Tessera
```

EnriqueL8 replied:
I need to circle back on this one based on the other conversation around the programming model for this

rodion-lim-partior replied:
Got it, Tessera public keys are heavily used for private networks, so it would be good if the data point is easily accessible

EnriqueL8 commented on internal/blockchain/ethereum/quorum/genesis.go:
```diff
@@ -0,0 +1,114 @@
+
+func CreateGenesis(addresses []string, blockPeriod int, chainID int64) *Genesis {
+	if blockPeriod == -1 {
+		blockPeriod = 0
```
Why did we go for a 0 second block period? If I recall correctly we set 5 as default for Besu

rodion-lim-partior replied:
This is the same Genesis configurations used for Geth. I kept it the same for consistency. For the purposes of a development environment, 5 seconds can be slightly too long since it slows down transactions and deployments

EnriqueL8 commented on internal/blockchain/ethereum/quorum/genesis.go:
```diff
@@ -0,0 +1,114 @@
+			},
+		},
+		Nonce:      "0x0",
+		Timestamp:  "0x60edb1c7",
```
I see we hardcode the timestamp and we have done the same for Besu in the pass, go quorum suggests using `0x0` https://docs.goquorum.consensys.io/reference/genesis

rodion-lim-partior replied:
Copied from Geth Genesis settings, this seems to be a timestamp 3 years in the past, which wouldn't cause any issues. But no harm changing to 0x0, since it's recommended in the Quorum docs as well.

EnriqueL8 commented on internal/blockchain/ethereum/quorum/geth_provider.go:
```diff
@@ -0,0 +1,409 @@
+	connector connector.Connector
+}
+
+func NewGethProvider(ctx context.Context, stack *types.Stack) *GethProvider {
```
Just a thought I know a lot of this code is inherited from the geth provider and that Quorum obviously includes a go-ethereum client but wondering if we should rename all of this to a `qurorum_provider` will leave up to you to think that through

EnriqueL8 replied:
The amount of shared code is quite large, it feels the only difference is the tesseraEnabled checks

rodion-lim-partior replied:
I struggled with this initially, but due to the other consensus algorithms and very different start up parameters for Quorum (e.g. needing to pass in other cli args, having to know the other Quorum endpoints, needing to wait for Tessera to be up first before starting the binary etc.), there are sufficiently large deviations from geth_provider.

Also, geth_provider in geth starts a single geth client, while geth_provider in quorum starts multiple quorum and or tessera pairs. Each quorum client maintains it's own private state trie, so we can't reuse the code to spin up a single client.

Once we add in support for IBFT/QBFT, there will be deviations in the `genesis.go` file in the same package as well. 

Will rename to quorum_provider instead of geth_provider

EnriqueL8 replied:
Thanks for the clarification

EnriqueL8 commented on internal/blockchain/ethereum/quorum/geth_provider.go:
```diff
@@ -0,0 +1,409 @@
+	connectorDependents := map[string]string{}
+	for i := 0; i < memberCount; i++ {
+		var dependsOn map[string]map[string]string
+		if p.stack.TesseraEnabled {
+			dependsOn = map[string]map[string]string{fmt.Sprintf("tessera_%d", i): {"condition": "service_started"}}
+		}
+		serviceDefinitions[i] = &docker.ServiceDefinition{
+			ServiceName: fmt.Sprintf("geth_%d", i),
+			Service: &docker.Service{
+				Image:         gethImage,
+				ContainerName: fmt.Sprintf("%s_geth_%d", p.stack.Name, i),
+				Volumes:       []string{fmt.Sprintf("geth_%d:/data", i)},
+				Logging:       docker.StandardLogOptions,
+				Ports:         []string{fmt.Sprintf("%d:8545", p.stack.ExposedBlockchainPort+(i*exposedBlockchainPortMultiplier))}, // defaults 5100, 5110, 5120, 5130
+				Environment:   p.stack.EnvironmentVars,
+				EntryPoint:    []string{"/bin/sh", "-c", "/data/docker-entrypoint.sh"},
+				DependsOn:     dependsOn,
+			},
+			VolumeNames: []string{fmt.Sprintf("geth_%d", i)},
+		}
+		connectorDependents[fmt.Sprintf("geth_%d", i)] = "service_started"
+
+		if p.stack.TesseraEnabled {
+			serviceDefinitions[i+memberCount] = &docker.ServiceDefinition{
```
nit move depends on lower
```suggestion
		serviceDefinitions[i] = &docker.ServiceDefinition{
			ServiceName: fmt.Sprintf("geth_%d", i),
			Service: &docker.Service{
				Image:         gethImage,
				ContainerName: fmt.Sprintf("%s_geth_%d", p.stack.Name, i),
				Volumes:       []string{fmt.Sprintf("geth_%d:/data", i)},
				Logging:       docker.StandardLogOptions,
				Ports:         []string{fmt.Sprintf("%d:8545", p.stack.ExposedBlockchainPort+(i*exposedBlockchainPortMultiplier))}, // defaults 5100, 5110, 5120, 5130
				Environment:   p.stack.EnvironmentVars,
				EntryPoint:    []string{"/bin/sh", "-c", "/data/docker-entrypoint.sh"},
				DependsOn:     dependsOn,
			},
			VolumeNames: []string{fmt.Sprintf("geth_%d", i)},
		}
		connectorDependents[fmt.Sprintf("geth_%d", i)] = "service_started"

		if p.stack.TesseraEnabled {
		        dependsOn = map[string]map[string]string{fmt.Sprintf("tessera_%d", i): {"condition": "service_started"}}
			serviceDefinitions[i+memberCount] = &docker.ServiceDefinition{
```

rodion-lim-partior replied:
quorum containers have a dependencies to tessera, shifted quorum docker service lower to avoid duplicating tesseraEnabled checks

EnriqueL8 commented on internal/blockchain/ethereum/quorum/geth_provider.go:
```diff
@@ -0,0 +1,409 @@
+		}
+	}
+
+	l.Info(fmt.Sprintf("generating quorum docker-entrypoint file for member %s", memberIndex))
+	quorumEntrypointOutputDirectory := filepath.Join(directory, "blockchain", fmt.Sprintf("geth_%s", memberIndex))
+	if err := CreateQuorumEntrypoint(p.ctx, quorumEntrypointOutputDirectory, gethVolumeName, memberIndex, p.stack.QuorumConsensus.String(), p.stack.Name, int(p.stack.ChainID()), p.stack.TesseraEnabled); err != nil {
+		return nil, err
+	}
```
The rename stuff specially when I see this

rodion-lim-partior replied:
Is this on geth being conflicting with quorum? If so, I renamed all geth references to quorum to make it clearer.

EnriqueL8 commented on internal/blockchain/ethereum/quorum/geth_provider.go:
```diff
@@ -0,0 +1,409 @@
+	if p.stack.TesseraEnabled {
+		tesseraVolumeName := fmt.Sprintf("%s_tessera_%s", p.stack.Name, memberIndex)
+		tesseraKeysOutputDirectory := filepath.Join(directory, "tessera", fmt.Sprintf("tessera_%s", memberIndex), "keystore")
+		_, tesseraPubKey, tesseraKeysPath, err = CreateTesseraKeys(p.ctx, tesseraImage, tesseraKeysOutputDirectory, "", "tm", keyPassword)
```
what does the name `tm` signify to a user for the new keys?

rodion-lim-partior replied:
tm signifies transaction manager. It's the default used in most of consensys's examples, ref: https://github.com/Consensys/quorum-dev-quickstart/blob/master/files/common/config/tessera/data/tessera-config-template.json

EnriqueL8 commented on internal/blockchain/ethereum/quorum/private_transaction_manager.go:
```diff
@@ -0,0 +1,169 @@
+	Data PrivateKeyData `json:"data"`
+}
+
+func CreateTesseraKeys(ctx context.Context, image, outputDirectory, prefix, name, password string) (privateKey, pubKey, path string, err error) {
```
I don't see you using the password

rodion-lim-partior replied:
good catch, removed

EnriqueL8 commented on internal/blockchain/ethereum/quorum/geth_provider.go:
```diff
@@ -0,0 +1,409 @@
+		l.Info(fmt.Sprintf("keys generated in %s", tesseraKeysPath))
+		l.Info(fmt.Sprintf("generating tessera docker-entrypoint file for member %s", memberIndex))
+		tesseraEntrypointOutputDirectory := filepath.Join(directory, "tessera", fmt.Sprintf("tessera_%s", memberIndex))
+		if err := CreateTesseraEntrypoint(p.ctx, tesseraEntrypointOutputDirectory, tesseraVolumeName, memberCount, p.stack.Name); err != nil {
```
I get that the TesseraEntrypoint is quite tied to the creation of the account with tessera keys but the CLI has followed a pattern where the blockchain configuration is stored in the `stack` directory such as the genesis and then at `ff start` it gets copied over so we don't envision anyone modifying this Tessera file and then restarting the stack? It won't get copied over again based on where this code is

rodion-lim-partior replied:
Refactored for docker entry point files to follow the same CLI pattern of storing them in the `stack` directory. I got pretty confused over here, are you saying if we follow the CLI pattern, we are able to modify the docker-entrypoint file and use the new file whenever we stop and start the stack?

EnriqueL8 commented on internal/blockchain/ethereum/quorum/quorum.go:
```diff
@@ -0,0 +1,112 @@
+if [ "istanbul" == "$GOQUORUM_CONS_ALGO" ];
+then
+    echo "Using istanbul for consensus algorithm..."
+    export CONSENSUS_ARGS="--istanbul.blockperiod 5 --mine --miner.threads 1 --miner.gasprice 0 --emitcheckpoints"
```
uhmm this block period is not using the configuration... I assume this overwrites the one in the genesis

rodion-lim-partior replied:
my bad, going to pass it in as a parameter into the function. Clique and QBFT takes from Genesis file, other options allows CLI flags

EnriqueL8 commented on internal/blockchain/ethereum/quorum/quorum.go:
```diff
@@ -0,0 +1,112 @@
+then
+    echo "Using qbft for consensus algorithm..."
+    export CONSENSUS_ARGS="--mine --miner.threads 1 --miner.gasprice 0 --emitcheckpoints"
+    export QUORUM_API="istanbul"
```
I know we are not supporting this for now and not completely familiar but did you mean to set `istanbul`?

rodion-lim-partior replied:
istanbul is the right API endpoint to expose, both QBFT and IBFT implementations inherits from the same base Istanbul package, but they share the same API group

EnriqueL8 reviewed (commented):
Thanks @rodion-lim-partior, great contribution! I spent some time going over the lower level details as well as reviewing Peter's high level architecture decisions. 

I realise the test coverage has significantly dropped in this repo in the past and was wondering if you could add test coverage for your bits as much as you can without significant refactoring and then I will raise a follow up issue to clean the debt that was caused my previous contributions!


Finally, I do not want to miss any future work mentioned as part of this PR so could you please raise the changed needed for these as issues to this repo please:

> Unsupported Consensus Mechanism (only when Quorum is selected)
> RAFT (Intend to support in future, needs command line tweaks, different discovery mechanism)
> IBFT (Intend to support in future, needs Genesis file tweaks)
> QBFT (Intend to support in future, needs Genesis file tweaks)

and 

>  Unsupported Client Combinations
> 
> Besu + EVMConnect + Tessera (Intend to support in future)

EnriqueL8 commented:
A few lint errors to fix as well:
```
Error: internal/blockchain/ethereum/quorum/private_transaction_manager.go:157:31: Error return value is not checked (errcheck)
	CopyTesseraEntrypointToVolume(ctx, outputDirectory, volumeName)
	                             ^
Error: internal/blockchain/ethereum/quorum/quorum.go:103:30: Error return value is not checked (errcheck)
	CopyQuorumEntrypointToVolume(ctx, outputDirectory, volumeName)
	                            ^
Error: internal/blockchain/ethereum/quorum/private_transaction_manager.go:76:43: unslice: could simplify pubKeyBytes[:] to pubKeyBytes (gocritic)
	return privateKeyData.Data.Bytes, string(pubKeyBytes[:]), path, nil
	                                         ^
Error: internal/blockchain/ethereum/quorum/quorum.go:28:121: ST1003: func parameter chainId should be chainID (stylecheck)
func CreateQuorumEntrypoint(ctx context.Context, outputDirectory, volumeName, memberIndex, consensus, stackName string, chainId int, tesseraEnabled bool) error {
                                                                                                                        ^
make: *** [Makefile:32: lint] Error 1
```

rodion-lim-partior commented:
@EnriqueL8, addressed your comments and refactored most of the code accordingly. Will add additional tests for my bits in the next commit. Also raised two more issues on the remaining bits of the PR.

EnriqueL8 commented:
Thanks, will give it another pass

EnriqueL8 reviewed (commented):
Thanks @rodion-lim-partior this is looking great and for following up with the issues! I'll like @peterbroadhurst to give this another pass now

rodion-lim-partior commented:
@EnriqueL8, added more tests to increase the coverage. Had to do some refactoring on the non quorum packages to allow for that.

peterbroadhurst commented:
@rodion-lim-partior - would like if it's ok to check these items, as I think this will need to go into the docs and I need a little more help understanding the options you're proposing:

> Quorum + EvmConnect + Tessera
> Quorum + EthConnect + Tessera (Haven't fully tested this, still undecided if this should be supported since we noticed the redirection in EthConnect repository to go with EvmConnect instead)

I'm not sure I understand how this fits into the first delivery here for a dev environment.
The code that makes Tessera private smart contracts work with FireFly, is only in EthConnect (not in EVMConnect):
https://github.com/hyperledger/firefly-ethconnect/blob/d3d27b26f0344dcb76f59a043523ae17d3f2141b/internal/eth/send.go#L225-L242

Wouldn't a "Quorum + EvmConnect + Tessera" stack be non-functional for users?

No problem for your PR to enable only one combination, but it seems to me like it should be "Quorum + EthConnect + Tessera".

> Quorum + EvmConnect (There is no strong use case to use Quorum without Tessera, we recommend going directly with Geth / Besu instead even though this is supported)
> Quorum + EthConnect (Same as above)

I understand that the future of the Quorum codebase is in-flux at the moment, but I do know from experience that the adoption of Quorum _without Tessera_ (or Constellation) is very significant for permissioned PoA (IBFT etc.) based networks. The momentum has certainly shifted very significantly to Hyperledger Besu for new networks.

The argument that Geth is in the box is valid so there's `go-ethereum` based runtime support, although that's also in even more doubt from a community perspective as Clique support is being removed from Geth.

In net - I'm very happy with your contribution not supporting Quorum+EvmConnect if you're not able to take on that work, but if you do so I think I'd likely go ahead and propose a follow-on commit before we include this in the same release to avoid needing to document why it's not supported.

rodion-lim-partior commented:
@peterbroadhurst, very valid comments, thank you. `EVMConnect` is something I am currently working on to bring Quorum / Tessera support. I intend to raise the PRs to the other repositories (firefly-signer, firefly-transaction-manager and firefly-evmconnect). As you mentioned, this includes back-porting some of the work that is already in EthConnect. 

I'm happy to change the wording to add future support for Quorum + EvmConnect / EthConnect, without a Tessera instance, since the use case you mentioned in your previous comment sounds valid.

It will be great if this PR can be merged before I attempt on the remaining ones to bring Quorum/Tessera support to the entire Firefly stack. Let me know your proposal on the follow-on commit, I'm happy to work on it.

peterbroadhurst commented:
Sounds like @rodion-lim-partior if the change you could make, is to disable the `EVMConnect` support in this PR (which could then go in quickly), and have a follow-on to turn it on once we've worked through the considerations in EVMConnect.

I'll then work on getting Quorum only support in, as that should be straight forward.

I think there will be some architectural complexity there, so if maybe you could start with an architecture proposal in FFTM for how you are thinking about it, that would be helpful. FFTM is blockchain agnostic, so we need to try and preserve that.

Doing a detail PR review of the other items now.

peterbroadhurst reviewed (changes requested):
Think the only required changes for this to merge are covered in this review:
- Tweaks to spelling of args
- Only allowing `evmconnect` until work is done elsewhere to support it in the framework

With those two small tweaks, very happy to see this go in. I'll look to help with the making of tessera optional.

rodion-lim-partior commented:
@peterbroadhurst, made the init parameters generic. Also disabled `evmconnect` support temporarily. I'm still looking through the transaction-manager and signer implementation to understand the extent of the changes required to bring Quorum specific support to evmconnect, will get back on the proposal in a separate issue / PR and see if it's possible to make it Blockchain agnostic

EnriqueL8 commented:
Thanks @rodion-lim-partior

peterbroadhurst reviewed (approved):
Thanks @rodion-lim-partior - great to see this go in.
